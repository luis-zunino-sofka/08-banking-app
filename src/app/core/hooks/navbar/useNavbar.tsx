import { ROUTE_PATH } from "@core/constants";
import { INavItem } from "@core/interfaces";
import { useAppContext } from "@core/state";
import { resetLoginData } from "@core/state/login";
import { useNavigate } from "react-router-dom";

export const useNavbar = () => {
  const navigate = useNavigate();
  const { dispatch } = useAppContext();

  const isActive = (path: string) => path === location.pathname;

  const handleLogout = () => {
    localStorage.clear();
    dispatch(resetLoginData(undefined));
    navigate(ROUTE_PATH.LOGIN);
  };

  const navItems: INavItem[] = [
    {
      path: ROUTE_PATH.DASHBOARD,
      label: "DepÃ³sito",
      onClick: () => navigate(ROUTE_PATH.DASHBOARD),
      isActive: isActive(ROUTE_PATH.DASHBOARD),
    },
    {
      path: ROUTE_PATH.PURCHASE,
      label: "Comprar",
      onClick: () => navigate(ROUTE_PATH.PURCHASE),
      isActive: isActive(ROUTE_PATH.PURCHASE),
    },
    {
      path: ROUTE_PATH.ACOUNT,
      label: "Gestion de cuentas",
      onClick: () => navigate(ROUTE_PATH.ACOUNT),
      isActive: isActive(ROUTE_PATH.ACOUNT),
    },
    {
      path: ROUTE_PATH.WITHDRAW,
      label: "Retirar",
      onClick: () => navigate(ROUTE_PATH.WITHDRAW),
      isActive: isActive(ROUTE_PATH.WITHDRAW),
    },
  ];

  return { navItems, handleLogout };
};

export default useNavbar;
