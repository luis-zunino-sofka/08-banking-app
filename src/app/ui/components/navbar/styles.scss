.navbar {
  display: flex;
  justify-content: space-between;
  align-items: center;
  background-color: #2d8c2f4d;
  padding: 5px 50px;
  color: #2d8c2f;
  top: 0;
  left: 0;
  width: 100%;
  z-index: 1000;
  margin-bottom: 4rem;
  &__title {
    font-size: 1rem;
    font-weight: bold;
  }

  &__balance {
    font-size: 1rem;

    strong {
      font-weight: bold;
      margin-left: 5px;
    }

    button.navbar__reload {
      background-color: #ffffff;
      color: #2d8c2f;
      font-size: 14px;
      border: none;
      padding: 5px 10px;
      cursor: pointer;
      border-radius: 5px;
      margin-left: 10px;

      &:hover {
        background-color: #f1f1f1;
      }
    }
  }
  &__nav {
    display: flex;
    justify-content: center;
  }
  &__list {
    display: flex;
    gap: 5px;
    list-style: none;
  }

  &__item {
    list-style: none;
  }

  &__link {
    padding: 5px;
    background-color: #2d8c2f4d;
    color: white;
    border-radius: 5px;
    text-align: center;
    text-decoration: none;
    font-weight: 500;
    font-size: 16px;
    cursor: pointer;
    transition: background-color 0.3s ease, transform 0.2s ease;

    &:hover {
      background-color: #2d8c2f;
      transform: translateY(-2px);
    }

    &:focus {
      outline: none;
    }
    &--active {
      background-color: #2d8c2f;
      transform: translateY(-2px);
    }
  }

  &__logout {
    background-color: #ff5722;
    color: white;
    border-radius: 5px;
    text-align: center;

    button {
      font-size: 16px;
      background-color: #ff5722;
      color: white;
      border: none;
      border-radius: 5px;
      cursor: pointer;

      &:hover {
        background-color: #e64a19;
      }
    }
  }
}
