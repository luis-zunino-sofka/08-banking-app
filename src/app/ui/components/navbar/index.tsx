import "./styles.scss";
import { INavItem, IState } from "@core/interfaces";
import { FC } from "react";

export interface IDashboardProps {
  state: IState;
  isRefechingAccounts: boolean;
  navItems: INavItem[];
  refetchAccounts: () => void;
  handleLogout: () => void;
}

export const Navbar: FC<IDashboardProps> = (props) => {
  const {
    state,
    navItems,
    isRefechingAccounts,
    refetchAccounts,
    handleLogout,
  } = props;

  return (
    <nav className="navbar">
      <div className="navbar__title">
        <h1>Bienvenido {state.loginData?.username}</h1>
      </div>
      <div className="navbar__balance">
        <span>Saldo disponible: </span>
        <strong>${state.balance}</strong>
        <button
          disabled={isRefechingAccounts}
          className="navbar__reload"
          onClick={() => refetchAccounts()}
        >
          Actualizar
        </button>
      </div>
      <div className="navbar__nav">
        <ul className="navbar__list">
          {navItems.map((item) => (
            <li key={item.path} className="navbar-item">
              <button
                className={`navbar__link ${
                  item.isActive ? "navbar__link--active" : ""
                }`}
                onClick={item.onClick}
              >
                {item.label}
              </button>
            </li>
          ))}
          <li className="navbar__item">
            <button
              className="navbar__link navbar__logout"
              onClick={handleLogout}
            >
              Cerrar sesi√≥n
            </button>
          </li>
        </ul>
      </div>
    </nav>
  );
};
