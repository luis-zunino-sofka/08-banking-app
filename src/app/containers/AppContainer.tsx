import { useNavbar } from "@core/hooks";
import { useAccountContext, useAppContext } from "@core/state";
import { Navbar } from "@ui/components/navbar";
import { AppLayout } from "@ui/layouts";
import React from "react";
import { Toaster } from "react-hot-toast";

export const AppContainer = ({ children }: { children: React.ReactNode }) => {
  const { state } = useAppContext();
  const { refetchAccounts } = useAccountContext();
  const { navItems, handleLogout } = useNavbar();

  return (
    <AppLayout>
      <Navbar
        state={state}
        handleLogout={handleLogout}
        refetchAccounts={refetchAccounts}
        navItems={navItems}
      />
      {children}
      <Toaster />
    </AppLayout>
  );
};
