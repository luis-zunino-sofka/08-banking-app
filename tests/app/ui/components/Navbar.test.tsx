import { render, screen, fireEvent } from "@testing-library/react";
import { useNavigate } from "react-router-dom";
import { beforeEach, describe, expect, test, vi } from "vitest";
import React from "react";
import { Navbar } from "../../../../src/app/ui/components/navbar/index";
import "@testing-library/jest-dom";
import { mockAccountState, mockAppState } from "../../../mocks/dataMocked";
import { AccountContextMock } from "../containers/AppContainer.test";
import useNavbar from "../../../../src/app/core/hooks/navbar/useNavbar";

vi.mock("../../../../src/app/core/hooks/navbar/useNavbar");
vi.mock("react-router-dom");

describe("Navbar", () => {
  const mockHandleLogout = vi.fn();
  const mockRefetchAccounts = vi.fn();
  const mockNavigate = vi.fn((value) => value);
  const mockedNavItems = [
    {
      path: "/dashboard",
      label: "Depósito",
      onClick: mockNavigate,
      isActive: false,
    },
    {
      path: "/purchase",
      label: "Comprar",
      onClick: mockNavigate,
      isActive: false,
    },
    {
      path: "/accounts",
      label: "Gestión de cuentas",
      onClick: mockNavigate,
      isActive: false,
    },
    {
      path: "/withdraw",
      label: "Retirar",
      onClick: mockNavigate,
      isActive: false,
    },
  ];

  beforeEach(() => {
    vi.mocked(useNavigate).mockReturnValue(mockNavigate);
    vi.mocked(useNavbar).mockReturnValue({
      navItems: mockedNavItems,
      handleLogout: mockHandleLogout,
    });
    render(
      <AccountContextMock
        mockAppState={mockAppState}
        children={
          <Navbar
            state={mockAppState}
            handleLogout={mockHandleLogout}
            refetchAccounts={mockRefetchAccounts}
            navItems={mockedNavItems}
          />
        }
        mockAccountState={mockAccountState}
      />
    );
  });

  test("Match snapshoot", () => {
    const { asFragment } = render(
      <AccountContextMock
        mockAppState={mockAppState}
        children={
          <Navbar
            state={mockAppState}
            handleLogout={mockHandleLogout}
            refetchAccounts={mockRefetchAccounts}
            navItems={mockedNavItems}
          />
        }
        mockAccountState={mockAccountState}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("renderiza correctamente el componente", () => {
    screen.getByText(/Bienvenido/i);
    screen.getByText(/\$0/i);
    screen.getByText(/Actualizar/i);
  });

  test('llama a handleLogout al hacer clic en "Cerrar sesión"', () => {
    const logoutButton = screen.getByText(/Cerrar sesión/i);
    fireEvent.click(logoutButton);

    expect(mockHandleLogout).toHaveBeenCalled();
  });

  test('llama a refetchAccounts al hacer clic en "Actualizar"', () => {
    const reloadButton = screen.getByText(/Actualizar/i);

    fireEvent.click(reloadButton);

    expect(mockRefetchAccounts).toHaveBeenCalled();
  });

  test("navega a la ruta correcta cuando se hace clic en los botones de navegación", () => {
    const dashboardButton = screen.getByText(/Depósito/i);
    const purchaseButton = screen.getByText(/Comprar/i);
    const accountButton = screen.getByText(/Gestión de cuentas/i);
    const withdrawButton = screen.getByText(/Retirar/i);

    fireEvent.click(dashboardButton);
    expect(mockNavigate).toHaveBeenCalled();

    fireEvent.click(purchaseButton);
    expect(mockNavigate).toHaveBeenCalled();

    fireEvent.click(accountButton);
    expect(mockNavigate).toHaveBeenCalled();

    fireEvent.click(withdrawButton);
    expect(mockNavigate).toHaveBeenCalled();
  });
});
