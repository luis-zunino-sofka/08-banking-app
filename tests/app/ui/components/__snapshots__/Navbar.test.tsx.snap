// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AccountProvider > Match snapshoot 1`] = `<DocumentFragment />`;

exports[`Navbar > Match snapshoot 1`] = `
<DocumentFragment>
  <nav
    class="navbar"
  >
    <div
      class="navbar__title"
    >
      <h1>
        Bienvenido 
      </h1>
    </div>
    <div
      class="navbar__balance"
    >
      <span>
        Saldo disponible: 
      </span>
      <strong>
        $0
      </strong>
      <button
        class="navbar__reload"
      >
        Actualizar
      </button>
    </div>
    <div
      class="navbar__nav"
    >
      <ul
        class="navbar__list"
      >
        <li
          class="navbar-item"
        >
          <button
            class="navbar__link "
          >
            Depósito
          </button>
        </li>
        <li
          class="navbar-item"
        >
          <button
            class="navbar__link "
          >
            Comprar
          </button>
        </li>
        <li
          class="navbar-item"
        >
          <button
            class="navbar__link "
          >
            Gestión de cuentas
          </button>
        </li>
        <li
          class="navbar-item"
        >
          <button
            class="navbar__link "
          >
            Retirar
          </button>
        </li>
        <li
          class="navbar__item"
        >
          <button
            class="navbar__link navbar__logout"
          >
            Cerrar sesión
          </button>
        </li>
      </ul>
    </div>
  </nav>
</DocumentFragment>
`;
